### PR TITLE
ARROW-6155: [Java] Extract a super interface for vectors whose elements reside in continuous memory segments

### PR DESCRIPTION
**Pull request by liyafan82**
For vectors whose data elements reside in continuous memory segments, they should implement a common super interface. This will avoid unnecessary code branches.

For now, such vectors include fixed-width vectors and variable-width vectors. In the future, there can be more vectors included.

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/FixedWidthVector.java
Comment (praveenbingo) [inline java/vector/src/main/java/org/apache/arrow/vector/FixedWidthVector.java]: please fix the java doc..

File: java/vector/src/main/java/org/apache/arrow/vector/ElementAddressableVector.java
Comment (praveenbingo) [inline java/vector/src/main/java/org/apache/arrow/vector/ElementAddressableVector.java]: should not all vectors have contiguous members? i think arrow buffers are designed for random access to any element.

do we need another interface or can we add them to value vectors.

File: java/vector/src/main/java/org/apache/arrow/vector/FixedWidthVector.java
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/FixedWidthVector.java]: nice catch. thank you @praveenbingo

File: java/vector/src/main/java/org/apache/arrow/vector/ElementAddressableVector.java
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/ElementAddressableVector.java]: It seems that not all vectors have continuous members.
For example, for a StructVector with multiple fields, its ith element is spread in multiple underlying vectors, which is not continuous.

Comment (praveenbingo) [reply java/vector/src/main/java/org/apache/arrow/vector/ElementAddressableVector.java]: ah i see what you mean, sounds good.

**Review by praveenbingo — APPROVED** [review]
+1. 

Thanks @liyafan82 Will merge by tomorrow if no one else has any comments.

**Comment by liyafan82** [comment]
> +1.
> 
> Thanks @liyafan82 Will merge by tomorrow if no one else has any comments.

Thank you @praveenbingo for your comments. 
Let's hold this until https://github.com/apache/arrow/pull/5020 is finished, to avoid extra merging effort for @emkornfield

**Comment by emkornfield** [comment]
Thanks @liyafan82  we shouldn't let #5020 stop from merging, I'll deal with the consequences of rebases.

**Comment by liyafan82** [comment]
> Thanks @liyafan82 we shouldn't let #5020 stop from merging, I'll deal with the consequences of rebases.

@emkornfield Don't worry. #5020 is a big change, and may involve substantial merging effort. I think we can at least hold current Java PRs for a few days.

**Comment by emkornfield** [comment]
Going to merge this.

**Pull request closed without merging**
